File: typescript/vault-rebalance/rebalance.ts
/**
 * Vault rebalance — swap tokens inside the vault without sending to anyone.
 *
 * Demonstrates Axon's executeSwap() endpoint. The bot signs a SwapIntent
 * specifying:
 *   - toToken: what to buy (e.g. WETH)
 *   - minToAmount: minimum output (slippage floor)
 *   - fromToken: what to sell (e.g. USDC) — bot-signed, prevents relayer substitution
 *   - maxFromAmount: max input (cost cap) — bot-signed, prevents overspending
 *
 * The relayer gets a Uniswap quote, verifies it's within the bot's signed
 * bounds, and executes on-chain. Output tokens stay in the vault.
 *
 * Use cases: DCA bots, yield rebalancing, portfolio management.
 *
 * Setup:
 *   1. Deploy vault on Base Sepolia, register bot
 *   2. Fund vault with USDC
 *   3. Set bot maxRebalanceAmount high enough (or 0 for no cap)
 *
 * Usage:
 *   cp .env.example .env
 *   npm install
 *   npx tsx rebalance.ts                     # swap 1 USDC -> WETH (default)
 *   npx tsx rebalance.ts 5                   # swap up to 5 USDC -> WETH
 *   npx tsx rebalance.ts balance             # check vault USDC + WETH balances
 */

import { AxonClient, Chain } from '@axonfi/sdk';
import { formatUnits } from 'viem';
import 'dotenv/config';

// ── Axon client ─────────────────────────────────────────────────────────────

const axon = new AxonClient({
  vaultAddress: process.env.AXON_VAULT_ADDRESS! as `0x${string}`,
  chainId: Number(process.env.AXON_CHAIN_ID ?? Chain.BaseSepolia),
  botPrivateKey: process.env.AXON_BOT_PRIVATE_KEY! as `0x${string}`,
});

// Base Sepolia token addresses
const USDC = '0x036CbD53842c5426634e7929541eC2318f3dCF7e' as const;
const WETH = '0x4200000000000000000000000000000000000006' as const;

// ── Commands ────────────────────────────────────────────────────────────────

async function cmdRebalance(maxUsdc: number) {
  console.log(`\nSwapping up to ${maxUsdc} USDC -> WETH (in-vault rebalance)`);
  console.log(`  fromToken:    USDC (bot-signed — relayer cannot change this)`);
  console.log(`  maxFromAmount: ${maxUsdc} USDC (bot-signed — caps input cost)`);
  console.log(`  toToken:      WETH`);

  // The bot signs: fromToken, maxFromAmount, toToken, minToAmount.
  // The relayer gets a Uniswap quote and verifies:
  //   - quote.fromAmount <= maxFromAmount (cost cap)
  //   - quote.toAmount >= minToAmount (slippage floor)
  // On-chain, the vault also checks maxRebalanceAmount per bot.
  // Calculate a reasonable minToAmount based on ~$2000/ETH with 10% slippage tolerance
  const estimatedEthPerUsdc = 1 / 2200; // conservative
  const minWeth = maxUsdc * estimatedEthPerUsdc * 0.9;
  // Convert to wei (18 decimals)
  const minWethWei = BigInt(Math.floor(minWeth * 1e18));

  console.log(`  minToAmount:  ${minWeth.toFixed(8)} WETH (${minWethWei} wei)\n`);

  let result = await axon.swap({
    toToken: 'WETH',
    minToAmount: minWethWei,
    fromToken: 'USDC',
    maxFromAmount: maxUsdc,
    memo: `Rebalance: up to ${maxUsdc} USDC -> WETH`,
  });

  // Poll if AI scan is triggered
  if (result.requestId && !result.txHash) {
    for (let i = 0; i < 30; i++) {
      result = await axon.pollSwap(result.requestId);
      if (result.status === 'approved' || result.status === 'rejected') break;
      console.log(`  status: ${result.status}...`);
      await new Promise((r) => setTimeout(r, 2000));
    }
  }

  console.log(`  Status: ${result.status}`);
  if (result.txHash) {
    console.log(`  TX: https://sepolia.basescan.org/tx/${result.txHash}`);
  }
  if (result.reason) {
    console.log(`  Reason: ${result.reason}`);
  }
}

async function cmdBalance() {
  const [usdc, weth] = await Promise.all([axon.getBalance(USDC), axon.getBalance(WETH)]);

  console.log(`\nVault balances:`);
  console.log(`  USDC:  ${formatUnits(usdc, 6)}`);
  console.log(`  WETH:  ${formatUnits(weth, 18)}`);
}

// ── CLI ─────────────────────────────────────────────────────────────────────

const USAGE = `
Vault rebalance — swap USDC -> WETH inside the vault

Usage:
  npx tsx rebalance.ts                     Swap up to 1 USDC -> WETH
  npx tsx rebalance.ts 5                   Swap up to 5 USDC -> WETH
  npx tsx rebalance.ts balance             Show vault balances
  npx tsx rebalance.ts help                Show this message
`;

async function main() {
  const cmd = process.argv[2] ?? '1';

  if (cmd === 'help') {
    console.log(USAGE);
    return;
  }
  if (cmd === 'balance') {
    console.log(`Vault: ${axon.vaultAddress}`);
    await cmdBalance();
    return;
  }

  console.log(`Vault: ${axon.vaultAddress}`);
  console.log(`Bot:   ${axon.botAddress}`);
  console.log(`Chain: Base Sepolia`);

  const amount = parseFloat(cmd);
  if (isNaN(amount)) {
    console.log(`Unknown command: ${cmd}`);
    console.log(USAGE);
    return;
  }

  await cmdBalance();
  await cmdRebalance(amount);
  await cmdBalance();
}

main().catch((err) => {
  console.error('Error:', err.message ?? err);
  process.exit(1);
});

File: python/vault-rebalance/rebalance.py
"""Vault rebalance — swap tokens inside the vault without sending to anyone.

Demonstrates Axon's executeSwap() endpoint. The bot signs a SwapIntent
specifying fromToken, maxFromAmount, toToken, and minToAmount. The relayer
gets a Uniswap quote, verifies it's within bounds, and executes on-chain.

Usage:
    pip install -r requirements.txt
    cp .env.example .env
    python rebalance.py                     # swap up to 1 USDC -> WETH
    python rebalance.py 5                   # swap up to 5 USDC -> WETH
    python rebalance.py balance             # check vault balances
"""

import asyncio
import os
import sys

from dotenv import load_dotenv

from axonfi import AxonClient, Chain

load_dotenv()

# Base Sepolia token addresses
USDC = "0x036CbD53842c5426634e7929541eC2318f3dCF7e"
WETH = "0x4200000000000000000000000000000000000006"


async def cmd_rebalance(client: AxonClient, max_usdc: float):
    print(f"\nSwapping up to {max_usdc} USDC -> WETH (in-vault rebalance)")
    print("  fromToken:     USDC (bot-signed — relayer cannot change this)")
    print(f"  maxFromAmount: {max_usdc} USDC (bot-signed — caps input cost)")
    print("  toToken:       WETH")
    # Calculate a reasonable minToAmount based on ~$2000/ETH with 10% slippage
    estimated_eth_per_usdc = 1 / 2200  # conservative
    min_weth = max_usdc * estimated_eth_per_usdc * 0.9
    min_weth_wei = int(min_weth * 1e18)
    print(f"  minToAmount:   {min_weth:.8f} WETH ({min_weth_wei} wei)\n")

    result = await client.swap(
        to_token="WETH",
        min_to_amount=min_weth_wei,
        from_token="USDC",
        max_from_amount=max_usdc,
        memo=f"Rebalance: up to {max_usdc} USDC -> WETH",
    )

    # Poll if AI scan triggered
    if result.request_id and not result.tx_hash:
        for _ in range(30):
            result = await client.poll_swap(result.request_id)
            if result.status in ("approved", "rejected"):
                break
            print(f"  status: {result.status}...")
            await asyncio.sleep(2)

    print(f"  Status: {result.status}")
    if result.tx_hash:
        print(f"  TX: https://sepolia.basescan.org/tx/{result.tx_hash}")
    if result.reason:
        print(f"  Reason: {result.reason}")


async def cmd_balance(client: AxonClient):
    usdc_bal = await client.get_balance(USDC)
    weth_bal = await client.get_balance(WETH)
    print("\nVault balances:")
    print(f"  USDC:  {int(usdc_bal) / 1e6:.6f}")
    print(f"  WETH:  {int(weth_bal) / 1e18:.18f}")


async def main():
    client = AxonClient(
        vault_address=os.environ["AXON_VAULT_ADDRESS"],
        chain_id=int(os.environ.get("AXON_CHAIN_ID", str(Chain.BaseSepolia))),
        bot_private_key=os.environ["AXON_BOT_PRIVATE_KEY"],
    )

    cmd = sys.argv[1] if len(sys.argv) > 1 else "1"

    if cmd == "help":
        print(__doc__)
        return

    if cmd == "balance":
        print(f"Vault: {client.vault_address}")
        await cmd_balance(client)
        return

    print(f"Vault: {client.vault_address}")
    print(f"Bot:   {client.bot_address}")
    print("Chain: Base Sepolia")

    amount = float(cmd)
    await cmd_balance(client)
    await cmd_rebalance(client, amount)
    await cmd_balance(client)


if __name__ == "__main__":
    asyncio.run(main())
